### ISSUE TITLE
Bug with multiple SpotLights in one Scene

### ISSUE THREAD

**Issue by C0br4** [Rajawali/Rajawali#1376]
I create multiple SpotLights in my Scene and update their position with setPosition() and the direction with setDirection().

The method setPostion() works as expected on each individual SpotLight. 
But when the setDirection() method is used, the direction of the first added SpotLight is applied to all the SpotLights in the Scene.

I'm using the current master branch of Rajawali.

**Comment by jwoolston** [comment]
I don't think I have observed this bug, but I also can't remember trying what you are saying. Im working on finishing switching everything to quaternions, so I'll make sure this is resolved in the process.

**Comment by ToxicBakery** [comment]
This may be a clone bug, are you cloning the first spot light to create the other spot lights?

**Comment by jwoolston** [comment]
@ToxicBakery that was my initial thought as well, but lights dont have a `clone()` method.

**Comment by C0br4** [comment]
I modified the SpotLight example from the RajawaliExamples to demonstrate the issue.

https://gist.github.com/C0br4/aab99aac3158af63d2aa

spotLight2 is always set to the same lookAt as the first spotLight

**Comment by jwoolston** [comment]
@C0br4 Thanks, that will help us figure out whats going on.

**Comment by jwoolston** [comment]
@C0br4 just wanted to give you an update on this issue. First of all, thankyou for raising it as it has uncovered a smattering of other issues with the library's internals so everything that extends from `ATransformable3D` will be the better for this fix. I am nearly done, just working out a few of the grander scale details.

**Comment by C0br4** [comment]
@jwoolston Thank you for taking the effort to fix this. I follow every commit on the issue-1376 branch and will report in if anything changed for the SpotLight lookat issue.

**Comment by jwoolston** [comment]
@C0br4 The major outstanding issue at present is that quaternion orientation has a degenerate case for looking at the same direction as the up vector. Turns out this is a huge issue with the library and I am focusing on fixing that before checking up on the multiple light issue. My suspicion is that the issue may disappear with the new implementation, and if it doesn't then I know it is a shader problem (we have been having some issues with the magic shader construction).

**Comment by jwoolston** [comment]
An update. I can confirm that this is not related to positioning or orientation of the light. I initialized a scene as follows:

```java
protected void initScene() {
			super.initScene();

			final SpotLight spotLight = new SpotLight();
			spotLight.setPower(1.5f);
            spotLight.enableLookAt();
            spotLight.setPosition(4.0, 4.0, 0.0);
            spotLight.setLookAt(0, 0, 0);
            spotLight.setColor(Color.RED);
            spot1 = spotLight;
			getCurrentScene().addLight(spotLight);

            final SpotLight spotLight2 = new SpotLight();
            spotLight2.setPower(1.5f);
            spotLight2.enableLookAt();
            spotLight2.setPosition(-4.0, 4.0, 0.0);
            spotLight2.setLookAt(0, 0, 0);
            spotLight2.setColor(Color.BLUE);
            spot2 = spotLight2;
            getCurrentScene().addLight(spotLight2);

           // Material code here - all the balls are white
}
```
The ball colors are all white to help show the lighting effects properly. We would expect from this initialization to see the right side illuminated red, the left side blue, and the middle a darkish purple.
After drawing the first frame, I log the position and direction vector of each light as follows:

```java
RajLog.d(this, "Red Light: " + spot1.getPosition());
RajLog.d(this, "Blue Light: " + spot2.getPosition());

RajLog.d(this, "Red Vector: " + spot1.getDirectionVector());
RajLog.d(this, "Blue Light: " + spot2.getDirectionVector());
```

The log produces the following output:

```
Red Light: Vector3 <x, y, z>: <4.0, 4.0, 0.0>
Blue Light: Vector3 <x, y, z>: <-4.0, 4.0, 0.0>
Red Vector: Vector3 <x, y, z>: <-0.7071067811865476, -0.7071067811865475, 3.3306690738754696E-16>
Blue Light: Vector3 <x, y, z>: <0.7071067811865476, -0.7071067811865475, 3.3306690738754696E-16>
```

As you can see, the positions are different, and the direction (look) vectors match what we would expect (45 degrees above the X axis in the XY plane on both sides of the Y axis). 

Unfortunately the corresponding visual is as follows:

![screenshot_2015-03-30-04-34-20](https://cloud.githubusercontent.com/assets/2993125/6890396/e7152644-d65e-11e4-8715-cd8b06c6d989.png)

So now I am investigating the inputs the the shader.

**Comment by jwoolston** [comment]
Shader indeed. The `applyParams()` method in `LightsVertexShaderFragment` has an issue when setting the light direction value for subsequent lights. You can see the line [here](https://github.com/Rajawali/Rajawali/blob/398299bb0593f1757ea8896946eb17dcb63caf82/rajawali/src/main/java/org/rajawali3d/materials/shaders/fragments/LightsVertexShaderFragment.java#L284). Specifically, the error is that it is trying to set the value of the uniform pointed to by -1, which is invalid.

**Comment by jwoolston** [comment]
Further inspection of the shaders performed by extracting the active uniform information showed that the subsequent light direction uniform was missing (as we expected based on the error). This lead me to take a more careful look at the shader code being compiled. In it I saw the following segment in the fragment shader:

```glsl
vec3 spotDir0 = normalize(uLightDirection0 * -1.0);
float spotFactor0 = dot(lightDir0, spotDir0);
if(uSpotCutoffAngle0<180.0) {
 		if(spotFactor0>=cos(radians(uSpotCutoffAngle0))) {
 			float exponent = 1.0 - cos(radians(uSpotCutoffAngle0));
 			exponent = 1.0 / exponent;
 			float facInv = 1.0 - spotFactor0;
 			exponent = facInv * exponent;
 			exponent = 1.0 - exponent;
 			spotFactor0 = pow(exponent, uSpotFalloff0 * 1.0 / exponent);
 		} else {
 			spotFactor0 = 0.0;
 		}
 		lightDir0 = vec3(lightDir0) * spotFactor0;
 }
vec3 lightDir1 = normalize(uLightPosition1 - vEye.xyz);
vec3 spotDir1 = normalize(uLightDirection0 * -1.0);
````

If you compare the first line and last line, you will notice that `uLightDirection0` is used for both `spotDir0` and `spotDir1`. This caused the shader compiler to optimize out the direction uniform for spot 1, resulting in the error above. Leading us to suspect the `LightsFragmentShaderFragment` class. If you look at the source [here](https://github.com/Rajawali/Rajawali/blob/398299bb0593f1757ea8896946eb17dcb63caf82/rajawali/src/main/java/org/rajawali3d/materials/shaders/fragments/LightsFragmentShaderFragment.java#L120) you can see that the `muLightDirection` array is being indexed by `lightAttCount` rather than `lightDirCount` as we would expect. Fixing this issue, we get our desired result: 

![screenshot_2015-03-30-07-18-40](https://cloud.githubusercontent.com/assets/2993125/6891737/88c2cdd0-d672-11e4-87fb-85ce2d1fdff0.png)

**Comment by jwoolston** [comment]
@C0br4 I feel pretty confident that this is now completely resolved. Once again, I'd like to extend my gratitude to you for reporting this bug. It turned out to be a tiny mistake in the code, but investigating it brought much more serious issues to my attention which I was able to resolve, greatly improving the robustness of the engine in the process. Additionally, it brought back many memories I had filed away of things which I wanted to take care of when I was last working on the engine, so all around good times! Happy coding! :tada:
